Add setting for users to determine frequency of backup creation

farewelltospring opened the pull request:
<!-- You can remove this first section if you have contributed before -->
### First time contributor checklist
<!-- replace the empty checkboxes [ ] below with checked ones [x] accordingly -->
- [x] I have read [how to contribute](https://github.com/signalapp/Signal-Android/blob/master/CONTRIBUTING.md) to this project
- [X] I have signed the [Contributor License Agreement](https://whispersystems.org/cla/)

### Contributor checklist
<!-- replace the empty checkboxes [ ] below with checked ones [x] accordingly -->
- [X] I am following the [Code Style Guidelines](https://github.com/signalapp/Signal-Android/wiki/Code-Style-Guidelines)
- [X] I have tested my contribution on these devices:
 * Sony Xperia 5iii, Android 13
 * Virtual device Pixel 3, Android 13
- [x] My contribution is fully baked and ready to be merged as is
- [X] I ensure that all the open issues my contribution fixes are mentioned in the commit message of my first commit using the `Fixes #1234` [syntax](https://help.github.com/articles/closing-issues-via-commit-messages/)

----------

### Description

This pull request allows users change the frequency that backups are made to a selected interval (weekly, monthly, quarterly, every 6 months, annually).

Currently, when backups are enabled, Signal will create a fresh backup every 24-hours. This causes unnecessary write cycles on the device's internal storage or SD card.

- Making backups is a power-intensive process which drains battery and uses CPU cycles which can put a strain on daily usage when users are out and about, and slow down other apps on the device while the backups are being created.
- Different users have different needs from backups. Some users message on Signal all day long, and others message on Signal infrequently. A 24-hour backup schedule is useful for the first group, but is of dubious value for the second group. It will write an identical file over and over.
- Low-end devices are wearing out prematurely. Many low-end Android phones have low-durability eMMC chips. A 24-hour backup schedule makes their already-short lifespan even shorter. eMMC chips will fail catastrophically when they reach end-of-life. The device bricks, and the user must replace the whole thing. **This disproportionately hurts Signal's poor users and produces environmental pollution.**

A setting to control the backup frequency would also give people more control over their data, which is an important value to cultivate in today's society.

Verified that this behaviour works by setting backup frequency to 30 days and then waiting that long to see if a backup got created (I restarted the device several times during that time period to verify that it persists).

https://github.com/signalapp/Signal-Android/assets/108375980/13f7643f-0d26-4ecb-b4a3-0b428c40ac07

Commit:
Show dialog to select backup frequency

Commit:
Add backup frequency as a setting that can be set

Commit:
todo commit

Commit:
Use frequency setting to schedule future backups

Commit:
Merge branch 'main' into slowbackups

Commit:
Pre-select backup frequency choice with user's current setting

Also minor Kotlin cleanups

Commit:
Rename backup frequency picker dialog class for accuracy

Commit:
Merge branch 'main' into slowbackups

Commit:
Show backup frequency in app settings

Yes I know this is bad practice but it's just for testing purposes

Commit:
Improve backup frequency picker buttons

Rename OK button and add cancel button

Commit:
Minor cleanups to backup frequency stuff

Commit:
Merge branch 'main' into slowbackups

Commit:
Add docstring for the unit of next backup time

Apparently it's a Unix timestamp in milliseconds and this took me way
too long to figure out myself

Commit:
Schedule next backup relative to last backup when changing frequency

When changing backup frequency settings, schedule the next backup using
the new setting, but relative to the timestamp of the latest backup,
instead of immediately.

This ensures that backups are not created too frequently. A backup will
only be immediately created if the latest backup was too long ago,
relative to the new frequency setting, or if there is no latest backup.

Commit:
Merge branch 'main' into slowbackups

Commit:
Convert raw strings to resources in backup settings

- Replace debug toast with debug log statement
- Use MaterialAlertDialog instead of boring AlertDialog

Commit:
Add newline to end of file

farewelltospring commented:
Fixed the localisation issue. I have only added a string for English but at least there are resources now.

Commit:
Convert string to plurals for non-English language

Just because the plural has a single manifestation in English doesn't
mean we don't need to have plurals for it in other languages.

Commit:
Merge branch 'main' into slowbackups

farewelltospring commented:
Is there anything I can do to make this PR better?

Commit:
Merge branch 'main' into slowbackups

Commit:
Merge branch 'main' into slowbackups

Commit:
Merge branch 'main' into slowbackups

LordBurtz commented:
@farewelltospring one suggestions, could you please add the option to put the number of days as an int (with sensible limits) as having to choose from a list feels arbitrary and might not cover all use cases, but I can totally understand if you keep it as is

+ i really like that feature

farewelltospring commented:
Just caught wind of this commit - https://github.com/signalapp/Signal-Android/commit/49ba83dda8ec7e9457dc620c960693f91590161e - is this PR able to be resolved?

Commit:
Merge branch 'main' into slowbackups

Commit:
Fix slow backups (farewelltospring's version)

There's some kind of divergence in the parallel implementation that the
people at Signal are doing instead of just merging my code (sadface),
which caused some lossy merge conflicts. This fixes it and restores my
version of slow backups to full functionality.

Commit:
Merge branch 'main' into slowbackups

Commit:
Merge branch 'main' into slowbackups

EmoBurrito commented:
Bumping because I'm genuinely surprised this functionality isn't present and I don't want this to be forgotten. I'm not copying my backup elsewhere daily, so there's no point in wasting power and slowing down my phone that frequently.

Commit:
Merge branch 'main' into slowbackups

tkrah commented:
Found this PR as *this* is the feature I really miss - what needs to be done to get this merged? This https://github.com/signalapp/Signal-Android/commit/49ba83dda8ec7e9457dc620c960693f91590161e is not be able to replace this PR because it is meant to be used in the upcoming remote chat backup - if you *WANT* a local backup, we still need this PR.

farewelltospring commented:
By "manual" you mean backups will only be created when the user presses the create button? I can get behind that.

Commit:
Merge branch 'main' into slowbackups

Commit:
Add a 'never' option for backup frequency, new UI

- Split frequency setting into separate setting entry in the UI
- Add 'never' option for backup frequency
- Do not schedule a backup when the user has selected 'never' for backup
  frequency
- Cleanups

farewelltospring commented:
Ok I uploaded some code. Full disclosure I haven't tested it yet because I gotta head out now and it'll take too long to compile, but I'll test it and take a video when I get back.

Commit:
Do not show frequency setting when backups are disabled

Also formatting

Commit:
More formatting :(

Commit:
Merge branch 'main' into slowbackups

Commit:
Add date label when backups are known to be enabled

Commit:
Merge branch 'main' into slowbackups

farewelltospring commented:
Demo recording attached.

https://github.com/user-attachments/assets/b953ddcd-a6d8-4a43-a658-236e1284acc1

Commit:
Rename backup frequency picker for consistency with backup time picker

farewelltospring commented:
Played around with it a little and I think found an edge case where if you change the frequency from a schedule to `NEVER`, then the original scheduled backup never gets canceled. So it'll create one more backup according to the previous frequency, and then stop afterwards. Do you think this is acceptable behaviour?

FWIW, I tried to dig into the code to figure out how to avoid this. There's a lot of abstraction and general cruft in the job management system (whose files are also not very well documented) and I can't immediately see a clean way to cancel a scheduled backup. **Could you give me a code pointer to cancel a scheduled job?** Or I can dig deeper myself at the cost of more time. Having said that, doesn't seem to persist across reboots, so there's that.

Curiously, one thing that *does* persist across reboots is in-progress backups. This seems to be a built-in feature to the JobRunner/JobDatabase/etc files, because logcat shows that it only tracks when the job was submitted and doesn't seem to respect the backup frequency setting when I switch it to `NEVER`.

IMO a good UX would be for switching the frequency setting to `NEVER` to:
1. Cancel future scheduled backups without relying on the user to reboot to clear it out
2. Cancel in-progress backups

Please let me know how you would like to move this PR forward.

farewelltospring commented:
Come through

alex-signal commented:
Cancellation methods are on JobManager, which you can access via `AppDependencies.getJobManager()`

Commit:
Cancel system alarm for backups when disabling backup frequency

This saves battery life by not requiring Android to wake up just to go
back to sleep after realising that backup frequency is disabled.

Commit:
Cancel ongoing backup jobs when the backup frequency is set to never

Commit:
Remove inaccurate/obsolete comment

Commit:
Rename function to apply new backup schedule setting for accuracy

Commit:
Merge branch 'main' into slowbackups

Commit:
Merge branch 'main' into slowbackups

farewelltospring commented:
:slightly_smiling_face:

farewelltospring commented:
🙂

farewelltospring commented:
@greyson-signal Is there anything I can do to make this easier for the team to review and merge?

Commit:
Merge branch 'main' into slowbackups

Commit:
Disable check to fix app

gabefair commented:
Hi, this fork is a bit behind main, could you update your branch?
@greyson-signal could this feature be merged?

Commit:
Merge branch 'main' into slowbackups

farewelltospring commented:
<img width="250" height="338" alt="image" src="https://github.com/user-attachments/assets/b995666d-16e5-40ea-8c9d-b3dccf258b60" />

tkrah commented:
> So we're actually in a weird spot now with adding something like this. We're in the midst of stabilizing our [new backup system](https://community.signalusers.org/t/public-signal-backups-testing/69984?u=greyson-signal), and with that comes a new local backup format that will likely solve every complaint you have in your initial post. Specifically, it'll be a differential backup, likely resulting in tens of MB's of data written per day (instead of GB's like it is now), and the entire backup will likely take seconds, not minutes.
> 
> For the new Signal-hosted backups, there's some important cryptography-related reasons why we'll be creating backups every day, and while those same reasons don't apply to the new local backups, I think it's gonna be weird if we allow scheduling for local and not Signal-hosted ones. And so I think for now I'm gonna officially "pause" this until we think about it some more.

Imho "we" thought about that more than 2 years now and I and I guess others too still miss this PR for my local backups. We need to setup + disable backups every week now for example and it is a hassle to do that, because I can't choose my passphrase, so every week I get a new one added to my list ... yeah.

Why it is weird to have that option only for scheduled backups, I don't see a problem with that difference - it is a different type of backup after all, local vs. signal hosted, so it can have different options imho.

+1 to move forward here and get this merged, please.

bmondream commented:
> I think it's gonna be weird if we allow scheduling for local and not Signal-hosted ones.

Scheduling could be added to local backups first and to Signal-hosted backups later. That's also an option.

I would like to see local backup scheduling implemented as well.

Commit:
Merge branch 'main' into slowbackups

Commit:
Merge branch 'main' into slowbackups

Commit:
Merge branch 'main' into slowbackups

bmondream commented:
I just tried it after figuring out how to build the project, and I noticed some to me unexpected behavior. As soon as I change the time of the scheduled backup, a backup is immediately created. I haven't yet tested if, in case of daily backups, it's created on time the next time (will report tomorrow).

Commit:
Merge branch 'main' into slowbackups

Commit:
Integrate v1 backup frequency into compose UI

Commit:
Merge branch 'main' into slowbackups
